File: backend/llm/equipment.go
package llm

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"regexp"

	"google.golang.org/genai"
)

// EquipmentItem is the structured output from LLM gear parsing.
type EquipmentItem struct {
	Name     string `json:"name"`
	Quantity int    `json:"quantity"`
	Notes    string `json:"notes"`
}

func ollamaURL() string {
	if u := os.Getenv("OLLAMA_URL"); u != "" {
		return u
	}
	return "http://host.docker.internal:11434"
}

func ollamaModel() string {
	if m := os.Getenv("OLLAMA_MODEL"); m != "" {
		return m
	}
	return "llama3.2"
}

func geminiModel() string {
	if m := os.Getenv("GEMINI_MODEL"); m != "" {
		return m
	}
	return "gemini-2.5-flash-lite"
}

var jsonArrayRe = regexp.MustCompile(`(?s)\[.*\]`)

func equipmentSystemPrompt() string {
	return `You are a mountain climbing equipment assistant. Parse the following gear description and return a JSON array. Each element must have exactly three fields:
	- "name": equipment name (string, in french)
	- "quantity": number needed (integer, 1 if unspecified)
	- "notes": "optional" or "mandatory" (translated in french), plus any relevant detail (string, in french)
	The name of these equipments are related with the mountain activities. You should only point out personal equipment, for instance quickdraws or rope.
	You should include only equipment you're absolutely sure about. Output ONLY the JSON array, no explanation.`
}

func equipmentUserPrompt(gearText string) string {
	return "Gear description:\n " + gearText
}

type Message struct {
    Role    string `json:"role"`
    Content string `json:"content"`
}

type OllamaRequest struct {
    Model    string    `json:"model"`
    Messages []Message `json:"messages"`
    Stream   bool      `json:"stream"`
    Options  map[string]any `json:"options"`
}

type OllamaResponse struct {
    Message Message `json:"message"`
}

// ExtractEquipment parses a free-form gear description into a structured list.
// lang is "fr" or "en" and drives the prompt language.
// Returns an error if Ollama is unreachable or returns unparseable output.
func ExtractEquipmentOllama(ctx context.Context, gearText, lang string) ([]EquipmentItem, error) {
	if gearText == "" {
		return nil, nil
	}

	reqBody := OllamaRequest{
        Model: ollamaModel(),
        Messages: []Message{
            {Role: "system", Content: equipmentSystemPrompt()},
            {Role: "user", Content: equipmentUserPrompt(gearText)},
        },
        Stream: false,
        Options: map[string]any{
            "num_predict": 512,
            "temperature": 0,   // greedy, deterministic JSON
        },
    }
	body, _ := json.Marshal(reqBody)
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, ollamaURL()+"/api/chat", bytes.NewReader(body))
	if err != nil {
		return nil, fmt.Errorf("building ollama request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("ollama unreachable: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("ollama returned status %d", resp.StatusCode)
	}

	raw, _ := io.ReadAll(resp.Body)
    var ollamaResp OllamaResponse
    if err := json.Unmarshal(raw, &ollamaResp); err != nil {
        return nil, err
    }
	
	var items []EquipmentItem
	if err := json.Unmarshal([]byte(ollamaResp.Message.Content), &items); err != nil {
		return nil, fmt.Errorf("parsing equipment JSON: %w", err)
	}

	return items, nil
}

func ExtractEquipmentGemini(ctx context.Context, gearText, lang string) ([]EquipmentItem, error) {
	// The client gets the API key from the environment variable `GEMINI_API_KEY`.
	client, err := genai.NewClient(ctx, nil)
	if err != nil {
		return nil, fmt.Errorf("creating Gemini client: %w", err)
	}

	result, err := client.Models.GenerateContent(
		ctx,
		geminiModel(),
		genai.Text(equipmentSystemPrompt()+"\n\n"+equipmentUserPrompt(gearText)),
		nil,
	)
	if err != nil {
		return nil, fmt.Errorf("calling Gemini API: %w", err)
	}

	resultStr := jsonArrayRe.FindString(result.Text())
	if resultStr == "" {
		return nil, fmt.Errorf("no JSON array found in Gemini response")
	}

	var items []EquipmentItem
	if err := json.Unmarshal([]byte(resultStr), &items); err != nil {
		return nil, fmt.Errorf("parsing equipment JSON: %w", err)
	}

	return items, nil
}

File: data/equipment.py
import json

from pydantic import BaseModel
from datasets import Dataset, DatasetDict, load_dataset
from typing import Optional, Self
from transformers import AutoTokenizer

SYSTEM_PROMPT = """You are a mountain climbing equipment assistant. Parse the following gear description and return a JSON array. Each element must have exactly three fields:
	- "name": equipment name (string, in french)
	- "quantity": number needed (integer, 1 if unspecified)
	- "notes": "optional" or "mandatory" (translated in french), plus any relevant detail (string, in french)
The name of these equipments are related with the mountain activities. You should only point out personal equipment, for instance quickdraws or rope.
You should include only equipment you're absolutely sure about. Output ONLY the JSON array, no explanation."""

tokenizer = AutoTokenizer.from_pretrained("meta-llama/Llama-3.2-3B-Instruct")

class Equipment(BaseModel):
    """
    An Equipment is a json array of equipement needed for a given route
    """

    id: int
    gear: str
    equipment: list[dict]
    prompt: Optional[str] = None
    completion: Optional[str] = None

    def make_prompt(self):

        messages = [
            {"role": "system", "content": SYSTEM_PROMPT},
            {"role": "user", "content": f"Gear description:\n {self.gear}"},
        ]

        self.prompt = tokenizer.apply_chat_template(
            messages,
            tokenize=False,
            add_generation_prompt=True  # adds the <|start_header_id|>assistant<|end_header_id|> opener
        )
        self.completion = json.dumps(self.equipment, ensure_ascii=False)

    def __repr__(self) -> str:
        return f"<{self.id} / Equipement: \n {json.dumps(self.equipment, indent=2, ensure_ascii=False)}\n>"

    def count_tokens(self, tokenizer):
        """Count tokens in the gear description"""
        return len(tokenizer.encode(self.prompt, add_special_tokens=False))
    @staticmethod
    def push_to_hub(dataset_name: str, train: list[Self], val: list[Self], test: list[Self]):
        """Push Item lists to HuggingFace Hub"""
        DatasetDict(
            {
                "train": Dataset.from_list([item.model_dump() for item in train]),
                "validation": Dataset.from_list([item.model_dump() for item in val]),
                "test": Dataset.from_list([item.model_dump() for item in test]),
            }
        ).push_to_hub(dataset_name)

    @classmethod
    def from_hub(cls, dataset_name: str) -> tuple[list[Self], list[Self], list[Self]]:
        """Load from HuggingFace Hub and reconstruct Items"""
        ds = load_dataset(dataset_name)
        return (
            [cls.model_validate(row) for row in ds["train"]],
            [cls.model_validate(row) for row in ds["validation"]],
            [cls.model_validate(row) for row in ds["test"]],
        )
    
    @staticmethod
    def parse(datapoint):
        return Equipment(
            id=datapoint["route_id"],
            gear=datapoint["gear"],
            equipment=datapoint["equipment"],
        )
